### DISCUSSION TITLE
How to collate list of graphs with same number of nodes but different number of features in pytorch geometric?

### DISCUSSION THREAD

**Discussion by anilkamat** [pyg-team/pytorch_geometric#5123] [Q&A]
Hi all, 
I am trying to create a PyG dataset using `DataLoader` from a list of graphs each having the same number of nodes but different number of features. Please see the snippet of the code below where the size of the window(data points) changes in every loop.  I also tried solving it with [pair graphs](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html#pairs-of-graphs) but it did not work. I am getting the following errors while collating. Any help or hit is greatly appreciated. 

RuntimeError: `  File "c:\Users\.....\dataloader.py", line 106, in process
    data, slices = self.collate(sequences)

  File "C:\Users\.....\Python39\lib\site-packages\torch_geometric\data\in_memory_dataset.py", line 111, in collate
    data, slices, _ = collate(
  File "C:\Users\....\Python39\lib\site-packages\torch_geometric\data\collate.py", line 84, in collate
    value, slices, incs = _collate(attr, values, data_list, stores,
  File "C:\Users\....\Python39\lib\site-packages\torch_geometric\data\collate.py", line 150, in _collate
    value = torch.cat(values, dim=cat_dim or 0, out=out)
RuntimeError: Sizes of tensors must match except in dimension 0. Expected size 40 but got size 90 for tensor number 1 in the list.`

a snippet of code: ` 
data = pd.read_csv(fileName, header=0, index_col=0).values

            config= {'HTA_LENGTH': [0,40,130,199,260,290,380] }
            sequences = []
            for i in range(5):
                g = Data()
                g.__num_nodes__ = n_node
                g.edge_index = edge_index
                window_start = config['HTA_LENGTH'][i]                    # starting timestamp of node features
                window_middle = config['HTA_LENGTH'][i+1]           # middle timestamp of node features
                window_end = config['HTA_LENGTH'][i+2]                # edn timestamp of node features
                full_window = np.swapaxes(data, 0, 1)   
                g.x = torch.FloatTensor(full_window[:, window_start:window_middle])   
                g.y = torch.FloatTensor(full_window[:, window_middle:window_end])   
                sequences += [g]
            data, slices = self.collate(sequences)
            torch.save((data, slices, n_node, mean, std_dev), self.processed_paths[j])`

**Comment by rusty1s** [comment]
In this case, I think you have to use padding, i.e. filling your features with zero and align their shapes.

**Reply by anilkamat** [reply]
Thank you for the response but I doubt this will not solve the problem. 
Since the difference between the length of the samples is in thousands, if I zero pad them they will lose the sequentially intact information which is not desirable for my correlation-based analysis. Is there any workaround to this issue?

**Reply by rusty1s** [reply]
How do you want to process these dynamically-shaped input features as part of your neural network? Instead of stacking them to a new tensor, you can also just hold them in a list. However, it will be harder to parallelize the compute that way.

**Reply by anilkamat** [reply]
Hi, Thank you. You are right! I used stacked the variable length input and output in a list and feed them to the model with the batch size of sample one.
